File: 86. Partition List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* partition(ListNode* head, int x) {
        ListNode* less = new ListNode(-1);
        ListNode* lesshead = less;
        ListNode* notless = new ListNode(-1);
        ListNode* notlesshead = notless;
        
        while (head != NULL)
        {
            if (head->val < x)
            {
                less->next = head;
                head = head->next;
                less = less->next;
                less->next = NULL;
            }
            else
            {
                notless->next = head;
                head = head->next;
                notless = notless->next;
                notless->next = NULL;
            }
        }
        less->next = notlesshead->next;
        return lesshead->next;
    }
};

File: 86. Partition List.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def partition(self, head, x):
        """
        :type head: ListNode
        :type x: int
        :rtype: ListNode
        """
        less = ListNode(-1)
        lesshead = less
        notless = ListNode(-1)
        notlesshead = notless
        while head:
            if head.val < x:
                less.next = head
                head = head.next
                less = less.next
                less.next = None
            elif head.val >= x:
                notless.next = head
                head = head.next
                notless = notless.next
                notless.next = None
        less.next = notlesshead.next
        return lesshead.next
